pkg/controller/delay.go

```diff
@@ -36,21 +36,24 @@ func cancelDelayedDownscaleIfConfigured(ctx context.Context, logger log.Logger,
 	callCancelDelayedDownscale(ctx, logger, httpClient, endpoints)
 }
 
-func checkScalingDelay(ctx context.Context, logger log.Logger, sts *v1.StatefulSet, httpClient httpClient, currentReplicas, desiredReplicas int32) error {
+// Checks if downscale delay has been reached on replicas in [desiredReplicas, currentReplicas) range.
+// If there is a range of replicas at the end of statefulset for which delay has been reached, this function
+// returns updated desired replicas that statefulset can be scaled to.
+func checkScalingDelay(ctx context.Context, logger log.Logger, sts *v1.StatefulSet, httpClient httpClient, currentReplicas, desiredReplicas int32) (updatedDesiredReplicas int32, _ error) {
 	if currentReplicas == desiredReplicas {
 		// should not happen
-		return nil
+		return currentReplicas, nil
 	}
 
 	delay, prepareURL, err := parseDelayedDownscaleAnnotations(sts.GetAnnotations())
 	if delay == 0 || prepareURL == nil || err != nil {
-		return err
+		return desiredReplicas, err
```

Comment:
If there was an error would be more correct to return `currentReplicas`. Today the caller doesn't scale to `updatedDesiredReplicas` in case of error, but I would like to protect from future bugs.

Reply:
>And follow-up fix in https://github.com/grafana/rollout-operator/commit/7e79c3b05f6c7a93442eba9bee65d17b4720cebd.

Exactly. That's what I had in mind.

Reply:
Fortunately test caught my mistake :)